### PR TITLE
OIDC Update & Change State API methods

### PR DESCRIPTION
**Pull request by talanknight**

### CODE REVIEW

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Looks great.  Just a few comments/suggestions.

File: internal/proto/local/controller/storage/auth/password/store/v1/password.proto [hunk 1]
```diff
@@ -21,11 +21,11 @@ message AuthMethod {
 
   // name is optional. If set, it must be unique within scope_id.
   // @inject_tag: `gorm:"default:null"`
-  string name = 4 [(custom_options.v1.mask_mapping) = {this:"name" that: "name"}];
+  string name = 4 [(custom_options.v1.mask_mapping) = {this:"Name" that: "name"}];
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/proto/local/controller/storage/auth/password/store/v1/password.proto:24]: It depends on how the repo implements the mask check.  The [pw auth method](https://github.com/hashicorp/boundary/blob/5162cc4570dfe19a74a4aa4f5f09f54b2225a595/internal/auth/password/repository_authmethod.go#L206) uses strings.EqualFold,  the oidc [auth method repo](https://github.com/hashicorp/boundary/blob/5162cc4570dfe19a74a4aa4f5f09f54b2225a595/internal/db/common/update.go#L147) uses checking map key presence checks.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 2]
```diff
@@ -237,6 +246,27 @@ func (s Service) UpdateAuthMethod(ctx context.Context, req *pbs.UpdateAuthMethod
 	return &pbs.UpdateAuthMethodResponse{Item: u}, nil
 }
 
+// ChangeState implements the interface pbs.AuthMethodServiceServer.
+func (s Service) ChangeState(ctx context.Context, req *pbs.ChangeStateRequest) (*pbs.ChangeStateResponse, error) {
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:309]: I've added nil checks in all validate* methods.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 3]
```diff
@@ -237,6 +246,27 @@ func (s Service) UpdateAuthMethod(ctx context.Context, req *pbs.UpdateAuthMethod
...
+	if err := validateChangeStateRequest(req); err != nil {
+		return nil, err
+	}
+	authResults := s.authResult(ctx, req.GetId(), action.ChangeState)
+	if authResults.Error != nil {
+		return nil, authResults.Error
+	}
+	am, err := s.changeStateInRepo(ctx, req.GetId(), req.GetState(), req.GetVersion())
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go]: Done.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 4]
```diff
@@ -237,6 +246,27 @@ func (s Service) UpdateAuthMethod(ctx context.Context, req *pbs.UpdateAuthMethod
...
+	if err != nil {
+		return nil, err
+	}
+	am.Scope = authResults.Scope
+	am.AuthorizedActions = authResults.FetchActionSetForId(ctx, am.Id, IdActions).Strings()
+	if err := populateCollectionAuthorizedActions(ctx, authResults, am); err != nil {
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:323]: Happy to chat.  This is just to populate the authorized actions on the resource returned.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 5]
```diff
@@ -379,16 +409,9 @@ func (s Service) listFromRepo(ctx context.Context, scopeIds []string) ([]*pb.Aut
 // createPwInRepo creates a password auth method in a repo and returns the result.
 // This method should never return a nil AuthMethod without returning an error.
 func (s Service) createPwInRepo(ctx context.Context, scopeId string, item *pb.AuthMethod) (*password.AuthMethod, error) {
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:470]: I've added nil checks in all validate* methods.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 6]
```diff
@@ -404,73 +427,9 @@ func (s Service) createPwInRepo(ctx context.Context, scopeId string, item *pb.Au
 // createOidcInRepo creates an oidc auth method in a repo and returns the result.
 // This method should never return a nil AuthMethod without returning an error.
 func (s Service) createOidcInRepo(ctx context.Context, scopeId string, item *pb.AuthMethod) (*oidc.AuthMethod, error) {
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:488]: I've added nil checks in all validate* methods which is done prior to this call.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 7]
```diff
@@ -508,37 +467,47 @@ func (s Service) createInRepo(ctx context.Context, scopeId string, item *pb.Auth
 	return toAuthMethodProto(out)
 }
 
-func (s Service) updateInRepo(ctx context.Context, scopeId, id string, mask []string, item *pb.AuthMethod) (*pb.AuthMethod, error) {
-	var opts []password.Option
-	if desc := item.GetDescription(); desc != nil {
-		opts = append(opts, password.WithDescription(desc.GetValue()))
+func (s Service) updateOidcInRepo(ctx context.Context, scopeId, id string, mask []string, item *pb.AuthMethod) (*oidc.AuthMethod, error) {
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:529]: I've added nil checks in all validate* methods.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 8]
```diff
@@ -550,24 +519,81 @@ func (s Service) updateInRepo(ctx context.Context, scopeId, id string, mask []st
...
 			return false, nil
 		}
-		return false, fmt.Errorf("unable to delete auth method: %w", err)
+		return false, fmt.Errorf("unable to delete auth method: %w", dErr)
 	}
 	return rows > 0, nil
 }
 
+func (s Service) changeStateInRepo(ctx context.Context, id, state string, version uint32) (*pb.AuthMethod, error) {
+	repo, err := s.oidcRepoFn()
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:636]: All invalid parameter checks is done in validateStateRequest which is performed prior to this method being called.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 9]
```diff
@@ -760,6 +786,104 @@ func toAuthTokenProto(t *authtoken.AuthToken) *pba.AuthToken {
 	}
 }
 
+func toStoragePwAuthMethod(scopeId string, item *pb.AuthMethod) (*password.AuthMethod, error) {
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:848]: I've added nil checks in all validate* methods.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 10]
```diff
@@ -851,17 +975,78 @@ func validateUpdateRequest(req *pbs.UpdateAuthMethodRequest) error {
 			if req.GetItem().GetType() != "" && auth.SubtypeFromType(req.GetItem().GetType()) != auth.PasswordSubtype {
 				badFields["type"] = "Cannot modify the resource type."
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go]: Extracted field names into consts as you suggested.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 11]
```diff
@@ -851,17 +975,78 @@ func validateUpdateRequest(req *pbs.UpdateAuthMethodRequest) error {
...
+						badFields["attributes.discovery_url"] = "The path should be empty or `/.well-known/openid-configuration`"
+					}
+				}
+			}
+			if handlers.MaskContains(req.GetUpdateMask().GetPaths(), "attributes.client_secret") && attrs.GetClientSecret().GetValue() == "" {
+				badFields["attributes.client_secret"] = "Cannot set the client secret to empty."
+			}
+
+			if handlers.MaskContains(req.GetUpdateMask().GetPaths(), "attributes.client_id") && attrs.GetClientId().GetValue() == "" {
+				badFields["attributes.client_id"] = "Can change but cannot unset this field."
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go]: I've changed them to make sure they are aligned.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 12]
```diff
@@ -883,6 +1068,25 @@ func validateListRequest(req *pbs.ListAuthMethodsRequest) error {
 	return nil
 }
 
+func validateChangeStateRequest(req *pbs.ChangeStateRequest) error {
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:1150]: I've added nil checks in all validate* methods.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 13]
```diff
@@ -883,6 +1068,25 @@ func validateListRequest(req *pbs.ListAuthMethodsRequest) error {
...
+	if st := auth.SubtypeFromId(req.GetId()); st != auth.OidcSubtype {
+		return handlers.NotFoundErrorf("This endpoint is only available for the %q Auth Method type.", auth.OidcSubtype.String())
+	}
+	badFields := make(map[string]string)
+	if req.GetVersion() == 0 {
+		badFields["version"] = "Resource version is required."
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go]: Made the field names consts.

File: internal/types/action/action.go [hunk 14]
```diff
@@ -39,6 +39,7 @@ const (
 	RemoveAccounts   Type = 30
 	ReadSelf         Type = 31
 	CancelSelf       Type = 32
+	ChangeState      Type = 33
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/types/action/action.go:42]: Woops, I missed the action.Map and action string method Thanks for catching that.

**Comment by jimlambrt** [comment]
Looks like there are some failing tests.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 15]
```diff
@@ -550,24 +519,81 @@ func (s Service) updateInRepo(ctx context.Context, scopeId, id string, mask []st
...
 	return rows > 0, nil
 }
 
+func (s Service) changeStateInRepo(ctx context.Context, id, state string, version uint32) (*pb.AuthMethod, error) {
+	repo, err := s.oidcRepoFn()
+	if err != nil {
+		return nil, err
+	}
+	var am *oidc.AuthMethod
+	switch state {
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go]: I've added one.  This is something that should have been caught in validateChangeStateRequest however.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 16]
```diff
@@ -883,6 +1068,25 @@ func validateListRequest(req *pbs.ListAuthMethodsRequest) error {
...
+func validateChangeStateRequest(req *pbs.ChangeStateRequest) error {
+	if st := auth.SubtypeFromId(req.GetId()); st != auth.OidcSubtype {
+		return handlers.NotFoundErrorf("This endpoint is only available for the %q Auth Method type.", auth.OidcSubtype.String())
+	}
+	badFields := make(map[string]string)
+	if req.GetVersion() == 0 {
+		badFields["version"] = "Resource version is required."
+	}
+	switch req.GetState() {
+	case stateInactive, statePrivate, statePublic:
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go]: Made the states an email.  ty!

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 17]
```diff
@@ -36,10 +37,15 @@ import (
 const (
 	loginNameKey = "login_name"
 	pwKey        = "password"
+
+	stateInactive = "inactive"
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go]: made the states an enum, ty!

File: internal/proto/local/controller/api/resources/authmethods/v1/auth_method.proto [hunk 18]
```diff
@@ -67,36 +67,36 @@ message OidcAuthMethodAttributes {
...
 	repeated string callback_urls = 90 [json_name="callback_urls"];
 
 	// certificates are optional PEM encoded x509 certificates that can be
 	// used as trust anchors when connecting to an OIDC provider.
-	repeated string certificates = 100 [json_name="certificates"];
+	repeated string certificates = 100 [json_name="certificates", (custom_options.v1.mask_mapping) = {this:"attributes.certificates" that: "Certificates"}];
 
 	// The audience claims for this oidc typed auth method.
-	repeated string audiences = 110 [json_name="audiences"];
+	repeated string audiences = 110 [json_name="audiences", (custom_options.v1.mask_mapping) = {this:"attributes.audiences" that: "AudClaims"}];
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/proto/local/controller/api/resources/authmethods/v1/auth_method.proto]: Renamed "allowed_audiences"

File: internal/proto/local/controller/storage/auth/oidc/store/v1/oidc.proto [hunk 19]
```diff
@@ -74,32 +74,32 @@ message AuthMethod {
...
   // @inject_tag: `gorm:"-"`
-  repeated string callback_urls = 170;
+  repeated string callback_urls = 170 [(custom_options.v1.mask_mapping) = {this:"CallbackUrls" that:"attributes.callback_url_prefixes"}];
 
   // aud_claims are the audience claims for a specific oidc auth method. These
   // are Value Objects that will be stored as AudClaim messages, and are
   // operatated on as a complete set.
   // @inject_tag: `gorm:"-"`
-  repeated string aud_claims = 180;
+  repeated string aud_claims = 180 [(custom_options.v1.mask_mapping) = {this:"AudClaims" that:"attributes.audiences"}];
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/proto/local/controller/storage/auth/oidc/store/v1/oidc.proto]: Renamed "allowed_audiences"

File: internal/proto/local/controller/storage/auth/password/store/v1/password.proto (hunk 1 above)
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/proto/local/controller/storage/auth/password/store/v1/password.proto:24]: BTW, pw auth method also calls BuildUpdatePaths(...) a few lines further down from the line you referenced:
https://github.com/hashicorp/boundary/blob/5162cc4570dfe19a74a4aa4f5f09f54b2225a595/internal/auth/password/repository_authmethod.go#L215

With all that said, I think common.BuildUpdatePaths(...) is intended to be used for "all" updates, so it sort of sets the standard... but I'm not advocating some big refactor here.  Perhaps we need a backlog issue to make off of this consistent so it follows one pattern?

File: internal/proto/local/controller/storage/auth/oidc/store/v1/oidc.proto (hunk 19 above)
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/proto/local/controller/storage/auth/oidc/store/v1/oidc.proto]: ty!

File: internal/types/action/action.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/types/action/action.go:42]: I think, you might need to still add a test for ChangeState in `TestAction(...)`

File: internal/servers/controller/handlers/authmethods/authmethod_service.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:276]: ty!

File: internal/servers/controller/handlers/authmethods/authmethod_service.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go:290]: ah.

File: internal/servers/controller/handlers/authmethods/authmethod_service.go [hunk 20]
```diff
@@ -851,17 +975,78 @@ func validateUpdateRequest(req *pbs.UpdateAuthMethodRequest) error {
...
+			}
+
+			if attrs.GetClientSecretHmac() != "" {
+				badFields["attributes.client_secret_hmac"] = "Field is read only."
+			}
+			if attrs.GetState() != "" {
+				badFields["attributes.state"] = "Field is read only."
+			}
+			if len(attrs.GetCallbackUrls()) > 0 {
+				badFields["attributes.callback_urls"] = "Field is read only."
```
[the remark this answers is not in the document]
Comment (jimlambrt) [reply internal/servers/controller/handlers/authmethods/authmethod_service.go]: I think, it's fine given that it's changing anywho.  ty for clarifying.

**Review by jimlambrt — APPROVED** [review]
No need for me to re-review after you add the one missing test.  Nice work!

**Pull request merged**